File: mycollab-services-pro/src/main/java/com/mycollab/pro/module/project/service/ibatis/ProjectTemplateServiceImpl.java
package com.mycollab.pro.module.project.service.ibatis;

import com.mycollab.common.dao.OptionValMapper;
import com.mycollab.common.domain.OptionVal;
import com.mycollab.common.domain.OptionValExample;
import com.mycollab.core.MyCollabException;
import com.mycollab.db.arguments.BasicSearchRequest;
import com.mycollab.db.arguments.NumberSearchField;
import com.mycollab.db.arguments.SetSearchField;
import com.mycollab.module.project.ProjectMemberStatusConstants;
import com.mycollab.module.project.domain.*;
import com.mycollab.module.project.domain.criteria.*;
import com.mycollab.module.project.service.*;
import com.mycollab.module.tracker.dao.BugRelatedItemMapper;
import com.mycollab.module.tracker.domain.*;
import com.mycollab.module.tracker.domain.criteria.BugSearchCriteria;
import com.mycollab.module.tracker.domain.criteria.ComponentSearchCriteria;
import com.mycollab.module.tracker.domain.criteria.VersionSearchCriteria;
import com.mycollab.module.tracker.service.BugService;
import com.mycollab.module.tracker.service.ComponentService;
import com.mycollab.module.tracker.service.VersionService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * @author myCollab Ltd
 * @since 5.2.8
 */
@Service
public class ProjectTemplateServiceImpl implements ProjectTemplateService {
    private static final Logger LOG = LoggerFactory.getLogger(ProjectTemplateService.class);

    @Autowired
    private ProjectService projectService;

    @Autowired
    private ProjectRoleService projectRoleService;

    @Autowired
    private ProjectMemberService projectMemberService;

    @Autowired
    private ProjectTaskService projectTaskService;

    @Autowired
    private BugService bugService;

    @Autowired
    private BugRelatedItemMapper bugRelatedItemMapper;

    @Autowired
    private MilestoneService milestoneService;

    @Autowired
    private ComponentService componentService;

    @Autowired
    private VersionService versionService;

    @Autowired
    private RiskService riskService;

    @Autowired
    private MessageService messageService;

    @Autowired
    private OptionValMapper optionValMapper;

    @Override
    public Integer cloneProject(Integer projectId, String newPrjName, String newPrjKey, Integer sAccountId, String username) {
        SimpleProject project = projectService.findById(projectId, sAccountId);
        if (project != null) {
            LOG.info("Clone project info");
            project.setId(null);
            project.setName(newPrjName);
            project.setShortname(newPrjKey);
            Integer newProjectId = projectService.savePlainProject(project, username);
            cloneProjectOptions(projectId, newProjectId, sAccountId);
            Map<Integer, Integer> mapRoleIds = cloneProjectRoles(projectId, newProjectId, username, sAccountId);
            cloneProjectMembers(projectId, newProjectId, mapRoleIds, username);
            cloneProjectMessages(projectId, newProjectId, username);
            cloneProjectRisks(projectId, newProjectId, username);
            Map<Integer, Integer> milestoneMapIds = cloneProjectMilestone(projectId, newProjectId, username);
            cloneProjectTasks(projectId, newProjectId, milestoneMapIds, username);
            Map<Integer, Integer> componentMapIds = cloneProjectComponents(projectId, newProjectId, username, sAccountId);
            Map<Integer, Integer> versionMapIds = cloneProjectVersions(projectId, newProjectId, username, sAccountId);
            cloneProjectBugs(projectId, newProjectId, milestoneMapIds, componentMapIds, versionMapIds, username, sAccountId);

            return newProjectId;
        } else {
            throw new MyCollabException("Can not find project with id " + projectId);
        }
    }

    private void cloneProjectOptions(Integer projectId, Integer newProjectId, Integer sAccountId) {
        OptionValExample ex = new OptionValExample();
        ex.createCriteria().andIsdefaultEqualTo(false).andSaccountidEqualTo(sAccountId).andExtraidEqualTo(projectId);
        List<OptionVal> optionVals = optionValMapper.selectByExample(ex);
        for (OptionVal optionVal : optionVals) {
            optionVal.setId(null);
            optionVal.setExtraid(newProjectId);
            optionValMapper.insert(optionVal);
        }
    }

    private Map<Integer, Integer> cloneProjectRoles(Integer projectId, Integer newProjectId, String username, Integer sAccountId) {
        LOG.info("Clone project roles");
        Map<Integer, Integer> mapRoleIds = new HashMap<>();
        ProjectRoleSearchCriteria searchCriteria = new ProjectRoleSearchCriteria();
        searchCriteria.setProjectId(new NumberSearchField(projectId));
        List<SimpleProjectRole> roles = projectRoleService.findPageableListByCriteria(new BasicSearchRequest<>(searchCriteria));
        for (SimpleProjectRole role : roles) {
            role.setId(null);
            role.setProjectid(newProjectId);
            Integer newRoleId = projectRoleService.saveWithSession(role, username);
            projectRoleService.savePermission(projectId, newRoleId, role.getPermissionMap(), sAccountId);
            mapRoleIds.put(role.getId(), newRoleId);
        }
        return mapRoleIds;
    }

    private void cloneProjectTasks(Integer projectId, Integer newProjectId, Map<Integer, Integer> milestoneMapIds, String username) {
        LOG.info("Clone project tasks");
        Map<Integer, Integer> taskMapIds = new HashMap<>();
        TaskSearchCriteria searchCriteria = new TaskSearchCriteria();
        searchCriteria.setProjectId(NumberSearchField.and(projectId));
        List<SimpleTask> tasks = projectTaskService.findPageableListByCriteria(new BasicSearchRequest<>(searchCriteria));
        cloneProjectTasks(newProjectId, milestoneMapIds, taskMapIds, tasks, username);
    }

    private void cloneProjectTasks(Integer newProjectId, Map<Integer, Integer> milestoneMapIds,
                                   Map<Integer, Integer> taskMapIds, List<SimpleTask> tasks, String username) {
        List<SimpleTask> pendingTasks = new ArrayList<>();
        for (SimpleTask task : tasks) {
            Integer taskId = task.getId();
            Integer parentTaskId = task.getParenttaskid();
            if (parentTaskId == null) {
                task.setId(null);
                task.setMilestoneid(milestoneMapIds.get(task.getMilestoneid()));
                task.setProjectid(newProjectId);
                Integer newTaskId = projectTaskService.saveWithSession(task, username);
                taskMapIds.put(taskId, newTaskId);
            } else {
                Integer candidateParentTaskId = taskMapIds.get(parentTaskId);
                if (candidateParentTaskId != null) {
                    task.setId(null);
                    task.setProjectid(newProjectId);
                    task.setMilestoneid(milestoneMapIds.get(task.getMilestoneid()));
                    task.setParenttaskid(candidateParentTaskId);
                    Integer newTaskId = projectTaskService.saveWithSession(task, username);
                    taskMapIds.put(taskId, newTaskId);
                } else {
                    pendingTasks.add(task);
                }
            }
        }
        if (pendingTasks.size() > 0) {
            cloneProjectTasks(newProjectId, milestoneMapIds, taskMapIds, pendingTasks, username);
        }
    }

    private Map<Integer, Integer> cloneProjectVersions(Integer projectId, Integer newProjectId, String username, Integer sAccountId) {
        LOG.info("Clone project versions");
        Map<Integer, Integer> versionMapIds = new HashMap<>();
        VersionSearchCriteria searchCriteria = new VersionSearchCriteria();
        searchCriteria.setProjectId(NumberSearchField.and(projectId));
        List<SimpleVersion> versions = versionService.findPageableListByCriteria(new BasicSearchRequest<>(searchCriteria));
        for (SimpleVersion version : versions) {
            Integer versionId = version.getId();
            version.setId(null);
            version.setProjectid(newProjectId);
            Integer newVersionId = versionService.saveWithSession(version, username);
            versionMapIds.put(versionId, newVersionId);
        }
        return versionMapIds;
    }

    private Map<Integer, Integer> cloneProjectComponents(Integer projectId, Integer newProjectId, String username, Integer sAccountId) {
        LOG.info("Clone project components");
        Map<Integer, Integer> componentMapIds = new HashMap<>();
        ComponentSearchCriteria searchCriteria = new ComponentSearchCriteria();
        searchCriteria.setProjectId(NumberSearchField.and(projectId));
        List<SimpleComponent> components = componentService.findPageableListByCriteria(new BasicSearchRequest<>(searchCriteria));
        for (SimpleComponent component : components) {
            Integer componentId = component.getId();
            component.setId(null);
            component.setProjectid(newProjectId);
            Integer newComponentId = componentService.saveWithSession(component, username);
            componentMapIds.put(componentId, newComponentId);
        }
        return componentMapIds;
    }

    private void cloneProjectBugs(Integer projectId, Integer newProjectId, Map<Integer, Integer> milestoneMapIds,
                                  Map<Integer, Integer> componentMapIds, Map<Integer, Integer> versionMapIds,
                                  String username, Integer sAccountId) {
        LOG.info("Clone project bugs");
        BugSearchCriteria searchCriteria = new BugSearchCriteria();
        searchCriteria.setProjectId(NumberSearchField.and(projectId));
        List<SimpleBug> bugs = bugService.findPageableListByCriteria(new BasicSearchRequest<>(searchCriteria));
        for (SimpleBug bug : bugs) {
            bug.setId(null);
            bug.setProjectid(newProjectId);
            bug.setMilestoneid(milestoneMapIds.get(bug.getMilestoneid()));
            Integer newBugId = bugService.saveWithSession(bug, username);

            List<Version> affectedVersions = bug.getAffectedVersions();
            for (Version version : affectedVersions) {
                BugRelatedItem bugRelatedItem = new BugRelatedItem();
                bugRelatedItem.setBugid(newBugId);
                bugRelatedItem.setType(SimpleRelatedBug.AFFVERSION);
                bugRelatedItem.setTypeid(versionMapIds.get(version.getId()));
                bugRelatedItemMapper.insert(bugRelatedItem);
            }

            List<Version> fixedVersions = bug.getFixedVersions();
            for (Version version : fixedVersions) {
                BugRelatedItem bugRelatedItem = new BugRelatedItem();
                bugRelatedItem.setBugid(newBugId);
                bugRelatedItem.setType(SimpleRelatedBug.FIXVERSION);
                bugRelatedItem.setTypeid(versionMapIds.get(version.getId()));
                bugRelatedItemMapper.insert(bugRelatedItem);
            }

            List<Component> components = bug.getComponents();
            for (Component component : components) {
                BugRelatedItem bugRelatedItem = new BugRelatedItem();
                bugRelatedItem.setBugid(newBugId);
                bugRelatedItem.setType(SimpleRelatedBug.COMPONENT);
                bugRelatedItem.setTypeid(componentMapIds.get(component.getId()));
                bugRelatedItemMapper.insert(bugRelatedItem);
            }
        }
    }

    private void cloneProjectMembers(Integer projectId, Integer newProjectId, Map<Integer, Integer> mapRoleIds, String username) {
        LOG.info("Clone project members");
        ProjectMemberSearchCriteria searchCriteria = new ProjectMemberSearchCriteria();
        searchCriteria.setProjectId(new NumberSearchField(projectId));
        searchCriteria.setStatuses(new SetSearchField<>(ProjectMemberStatusConstants.ACTIVE,
                ProjectMemberStatusConstants.NOT_ACCESS_YET));
        List<SimpleProjectMember> members = projectMemberService.findPageableListByCriteria(new BasicSearchRequest<>(searchCriteria));
        for (SimpleProjectMember member : members) {
            member.setId(null);
            member.setProjectid(newProjectId);
            if (Boolean.FALSE.equals(member.getIsadmin())) {
                Integer newRoleId = mapRoleIds.get(member.getProjectroleid());
                member.setProjectroleid(newRoleId);
            }
            projectMemberService.saveWithSession(member, username);
        }
    }

    private void cloneProjectMessages(Integer projectId, Integer newProjectId, String username) {
        LOG.info("Clone project messages");
        MessageSearchCriteria searchCriteria = new MessageSearchCriteria();
        searchCriteria.setProjectids(new SetSearchField<>(projectId));
        List<SimpleMessage> messages = messageService.findPageableListByCriteria(new BasicSearchRequest<>
                (searchCriteria, 0, Integer.MAX_VALUE));
        for (SimpleMessage message : messages) {
            message.setId(null);
            message.setProjectid(newProjectId);
            messageService.saveWithSession(message, username);
        }
    }

    private void cloneProjectRisks(Integer projectId, Integer newProjectId, String username) {
        LOG.info("Clone project risks");
        RiskSearchCriteria searchCriteria = new RiskSearchCriteria();
        searchCriteria.setProjectId(NumberSearchField.and(projectId));
        List<SimpleRisk> risks = riskService.findPageableListByCriteria(new BasicSearchRequest<>(searchCriteria));
        for (SimpleRisk risk : risks) {
            risk.setId(null);
            risk.setProjectid(newProjectId);
            riskService.saveWithSession(risk, username);
        }
    }

    private Map<Integer, Integer> cloneProjectMilestone(Integer projectId, Integer newProjectId, String username) {
        LOG.info("Clone project milestones");
        Map<Integer, Integer> milestoneMapIds = new HashMap<>();
        MilestoneSearchCriteria searchCriteria = new MilestoneSearchCriteria();
        searchCriteria.setProjectIds(new SetSearchField<>(projectId));
        List<SimpleMilestone> milestones = milestoneService.findPageableListByCriteria(new BasicSearchRequest<>(searchCriteria));
        for (SimpleMilestone milestone : milestones) {
            Integer milestoneId = milestone.getId();
            milestone.setId(null);
            milestone.setProjectid(newProjectId);
            Integer newMilestoneId = milestoneService.saveWithSession(milestone, username);
            milestoneMapIds.put(milestoneId, newMilestoneId);
        }
        return milestoneMapIds;
    }
}

File: mycollab-services-pro/src/main/java/com/mycollab/pro/module/project/service/impl/ProjectTemplateServiceImpl.kt
package com.mycollab.pro.module.project.service.impl

import com.mycollab.common.dao.OptionValMapper
import com.mycollab.common.domain.OptionValExample
import com.mycollab.core.MyCollabException
import com.mycollab.db.arguments.BasicSearchRequest
import com.mycollab.db.arguments.NumberSearchField
import com.mycollab.db.arguments.SetSearchField
import com.mycollab.module.project.ProjectMemberStatusConstants
import com.mycollab.module.project.domain.*
import com.mycollab.module.project.domain.criteria.*
import com.mycollab.module.project.service.*
import com.mycollab.module.tracker.dao.BugRelatedItemMapper
import com.mycollab.module.tracker.domain.*
import com.mycollab.module.tracker.domain.criteria.BugSearchCriteria
import com.mycollab.module.tracker.domain.criteria.ComponentSearchCriteria
import com.mycollab.module.tracker.domain.criteria.VersionSearchCriteria
import com.mycollab.module.tracker.service.BugService
import com.mycollab.module.tracker.service.ComponentService
import com.mycollab.module.tracker.service.VersionService
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.util.*

/**
 * @author myCollab Ltd
 * @since 5.2.8
 */
@Service
class ProjectTemplateServiceImpl(private val projectService: ProjectService,
                                 private val projectRoleService: ProjectRoleService,
                                 private val projectMemberService: ProjectMemberService,
                                 private val projectTaskService: ProjectTaskService,
                                 private val bugService: BugService,
                                 private val bugRelatedItemMapper: BugRelatedItemMapper,
                                 private val milestoneService: MilestoneService,
                                 private val componentService: ComponentService,
                                 private val versionService: VersionService,
                                 private val riskService: RiskService,
                                 private val messageService: MessageService,
                                 private val optionValMapper: OptionValMapper) : ProjectTemplateService {

    override fun cloneProject(projectId: Int, newPrjName: String, newPrjKey: String, sAccountId: Int, username: String): Int? {
        val project = projectService.findById(projectId, sAccountId)
        if (project != null) {
            LOG.info("Clone project info")
            project.id = null
            project.name = newPrjName
            project.shortname = newPrjKey
            val newProjectId = projectService.savePlainProject(project, username)
            cloneProjectOptions(projectId, newProjectId, sAccountId)
            val mapRoleIds = cloneProjectRoles(projectId, newProjectId, username, sAccountId)
            cloneProjectMembers(projectId, newProjectId, mapRoleIds, username)
            cloneProjectMessages(projectId, newProjectId, username)
            cloneProjectRisks(projectId, newProjectId, username)
            val milestoneMapIds = cloneProjectMilestone(projectId, newProjectId, username)
            cloneProjectTasks(projectId, newProjectId, milestoneMapIds, username)
            val componentMapIds = cloneProjectComponents(projectId, newProjectId, username)
            val versionMapIds = cloneProjectVersions(projectId, newProjectId, username)
            cloneProjectBugs(projectId, newProjectId, milestoneMapIds, componentMapIds, versionMapIds, username)

            return newProjectId
        } else {
            throw MyCollabException("Can not find project with id $projectId")
        }
    }

    private fun cloneProjectOptions(projectId: Int?, newProjectId: Int?, sAccountId: Int?) {
        val ex = OptionValExample()
        ex.createCriteria().andIsdefaultEqualTo(false).andSaccountidEqualTo(sAccountId).andExtraidEqualTo(projectId)
        val optionVals = optionValMapper.selectByExample(ex)
        optionVals.forEach {
            it.id = null
            it.extraid = newProjectId
            optionValMapper.insert(it)
        }
    }

    private fun cloneProjectRoles(projectId: Int, newProjectId: Int?, username: String, sAccountId: Int): Map<Int, Int> {
        LOG.info("Clone project roles")
        val mapRoleIds = mutableMapOf<Int, Int>()
        val searchCriteria = ProjectRoleSearchCriteria()
        searchCriteria.projectId = NumberSearchField(projectId)
        val roles = projectRoleService.findPageableListByCriteria(BasicSearchRequest(searchCriteria)) as List<SimpleProjectRole>
        roles.forEach {
            it.id = null
            it.projectid = newProjectId
            val newRoleId = projectRoleService.saveWithSession(it, username)
            projectRoleService.savePermission(projectId, newRoleId, it.permissionMap, sAccountId)
            mapRoleIds.put(it.id, newRoleId)
        }
        return mapRoleIds
    }

    private fun cloneProjectTasks(projectId: Int, newProjectId: Int, milestoneMapIds: Map<Int, Int>, username: String) {
        LOG.info("Clone project tasks")
        val taskMapIds = HashMap<Int, Int>()
        val searchCriteria = TaskSearchCriteria()
        searchCriteria.projectId = NumberSearchField.equal(projectId)
        val tasks = projectTaskService.findPageableListByCriteria(BasicSearchRequest(searchCriteria)) as List<SimpleTask>
        cloneProjectTasks(newProjectId, milestoneMapIds, taskMapIds, tasks, username)
    }

    private fun cloneProjectTasks(newProjectId: Int?, milestoneMapIds: Map<Int, Int>,
                                  taskMapIds: MutableMap<Int, Int>, tasks: List<SimpleTask>, username: String) {
        val pendingTasks = ArrayList<SimpleTask>()
        tasks.forEach {
            val taskId = it.id
            val parentTaskId = it.parenttaskid
            if (parentTaskId == null) {
                it.id = null
                it.milestoneid = milestoneMapIds[it.milestoneid]
                it.projectid = newProjectId
                val newTaskId = projectTaskService.saveWithSession(it, username)
                taskMapIds.put(taskId, newTaskId)
            } else {
                val candidateParentTaskId = taskMapIds[parentTaskId]
                if (candidateParentTaskId != null) {
                    it.id = null
                    it.projectid = newProjectId
                    it.milestoneid = milestoneMapIds[it.milestoneid]
                    it.parenttaskid = candidateParentTaskId
                    val newTaskId = projectTaskService.saveWithSession(it, username)
                    taskMapIds.put(taskId, newTaskId)
                } else {
                    pendingTasks.add(it)
                }
            }
        }
        if (pendingTasks.size > 0) {
            cloneProjectTasks(newProjectId, milestoneMapIds, taskMapIds, pendingTasks, username)
        }
    }

    private fun cloneProjectVersions(projectId: Int, newProjectId: Int, username: String): Map<Int, Int> {
        LOG.info("Clone project versions")
        val versionMapIds = HashMap<Int, Int>()
        val searchCriteria = VersionSearchCriteria()
        searchCriteria.projectId = NumberSearchField.equal(projectId)
        val versions = versionService.findPageableListByCriteria(BasicSearchRequest(searchCriteria)) as List<SimpleVersion>
        versions.forEach {
            val versionId = it.id
            it.id = null
            it.projectid = newProjectId
            val newVersionId = versionService.saveWithSession(it, username)
            versionMapIds.put(versionId, newVersionId)
        }
        return versionMapIds
    }

    private fun cloneProjectComponents(projectId: Int, newProjectId: Int, username: String): Map<Int, Int> {
        LOG.info("Clone project components")
        val componentMapIds = HashMap<Int, Int>()
        val searchCriteria = ComponentSearchCriteria()
        searchCriteria.projectId = NumberSearchField.equal(projectId)
        val components = componentService.findPageableListByCriteria(BasicSearchRequest(searchCriteria)) as List<SimpleComponent>
        components.forEach {
            val componentId = it.id
            it.id = null
            it.projectid = newProjectId
            val newComponentId = componentService.saveWithSession(it, username)
            componentMapIds.put(componentId, newComponentId)
        }
        return componentMapIds
    }

    private fun cloneProjectBugs(projectId: Int, newProjectId: Int, milestoneMapIds: Map<Int, Int>,
                                 componentMapIds: Map<Int, Int>, versionMapIds: Map<Int, Int>,
                                 username: String) {
        LOG.info("Clone project bugs")
        val searchCriteria = BugSearchCriteria()
        searchCriteria.projectId = NumberSearchField.equal(projectId)
        val bugs = bugService.findPageableListByCriteria(BasicSearchRequest(searchCriteria)) as List<SimpleBug>
        bugs.forEach {
            it.id = null
            it.projectid = newProjectId
            it.milestoneid = milestoneMapIds[it.milestoneid]
            val newBugId = bugService.saveWithSession(it, username)

            val affectedVersions = it.affectedVersions
            affectedVersions?.forEach {
                val bugRelatedItem = BugRelatedItem()
                bugRelatedItem.bugid = newBugId
                bugRelatedItem.type = SimpleRelatedBug.AFFVERSION
                bugRelatedItem.typeid = versionMapIds[it.id]
                bugRelatedItemMapper.insert(bugRelatedItem)
            }

            val fixedVersions = it.fixedVersions
            fixedVersions?.forEach {
                val bugRelatedItem = BugRelatedItem()
                bugRelatedItem.bugid = newBugId
                bugRelatedItem.type = SimpleRelatedBug.FIXVERSION
                bugRelatedItem.typeid = versionMapIds[it.id]
                bugRelatedItemMapper.insert(bugRelatedItem)
            }

            val components = it.components
            components?.forEach {
                val bugRelatedItem = BugRelatedItem()
                bugRelatedItem.bugid = newBugId
                bugRelatedItem.type = SimpleRelatedBug.COMPONENT
                bugRelatedItem.typeid = componentMapIds[it.id]
                bugRelatedItemMapper.insert(bugRelatedItem)
            }
        }
    }

    private fun cloneProjectMembers(projectId: Int, newProjectId: Int?, mapRoleIds: Map<Int, Int>, username: String) {
        LOG.info("Clone project members")
        val searchCriteria = ProjectMemberSearchCriteria()
        searchCriteria.projectIds = SetSearchField(projectId)
        searchCriteria.statuses = SetSearchField(ProjectMemberStatusConstants.ACTIVE,
                ProjectMemberStatusConstants.NOT_ACCESS_YET)
        val members = projectMemberService.findPageableListByCriteria(BasicSearchRequest(searchCriteria)) as List<SimpleProjectMember>
        members.forEach {
            it.id = null
            it.projectid = newProjectId
            if (java.lang.Boolean.FALSE == it.isadmin) {
                val newRoleId = mapRoleIds[it.projectroleid]
                it.projectroleid = newRoleId
            }
            projectMemberService.saveWithSession(it, username)
        }
    }

    private fun cloneProjectMessages(projectId: Int, newProjectId: Int, username: String) {
        LOG.info("Clone project messages")
        val searchCriteria = MessageSearchCriteria()
        searchCriteria.projectids = SetSearchField(projectId)
        val messages = messageService.findPageableListByCriteria(BasicSearchRequest(searchCriteria, 0, Integer.MAX_VALUE)) as List<SimpleMessage>
        messages.forEach {
            it.id = null
            it.projectid = newProjectId
            messageService.saveWithSession(it, username)
        }
    }

    private fun cloneProjectRisks(projectId: Int, newProjectId: Int, username: String) {
        LOG.info("Clone project risks")
        val searchCriteria = RiskSearchCriteria()
        searchCriteria.projectId = NumberSearchField.equal(projectId)
        val risks = riskService.findPageableListByCriteria(BasicSearchRequest(searchCriteria)) as List<SimpleRisk>
        risks.forEach {
            it.id = null
            it.projectid = newProjectId
            riskService.saveWithSession(it, username)
        }
    }

    private fun cloneProjectMilestone(projectId: Int, newProjectId: Int, username: String): Map<Int, Int> {
        LOG.info("Clone project milestones")
        val milestoneMapIds = HashMap<Int, Int>()
        val searchCriteria = MilestoneSearchCriteria()
        searchCriteria.projectIds = SetSearchField(projectId)
        val milestones = milestoneService.findPageableListByCriteria(BasicSearchRequest(searchCriteria)) as List<SimpleMilestone>
        milestones.forEach {
            val milestoneId = it.id
            it.id = null
            it.projectid = newProjectId
            val newMilestoneId = milestoneService.saveWithSession(it, username)
            milestoneMapIds.put(milestoneId, newMilestoneId)
        }
        return milestoneMapIds
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(ProjectTemplateService::class.java)
    }
}
